[PP] Add eval() API to schedule

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #157795

These change add an `eval()` API to PP schedules

## Context

Currently, you can run "Forward only" for a schedule in two ways:
1. Use a custom schedule `_ScheduleForwardOnly`
2. Do not pass in `loss_fn` in schedule constructor, and no backward computations will be executed.

However, this is still limiting because we may want to run forward through the pipeline / calculate the loss, but without backward, e.g. during validation. These changes allow for this.

```python
if self.rank == 0:
    schedule.eval(x)
elif self.rank == self.world_size - 1:
    losses = []
    schedule.eval(target=target, losses=losses)
else:
    schedule.eval()
```

TODO:
- in later PRs, we will deprecate the `_ScheduleForwardOnly`

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

H-Huang committed:
[PP] Allow schedules to run under torch.no_grad()

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] Allow schedules to run under torch.no_grad()"


This change allows PP schedules to run under `with torch.no_grad():`

## Context

Currently, you can run "Forward only" for a schedule in two ways:
1. Use a custom schedule `_ScheduleForwardOnly`
2. Do not pass in `loss_fn` in schedule constructor, and no backward computations will be executed.

However, this is still limiting because we may want to run forward through the pipeline / calculate the loss, but without backward, e.g. during validation. These changes allow for this.

```python
with torch.no_grad():
    if self.rank == 0:
        schedule.step(x)
    elif self.rank == self.world_size - 1:
        losses = []
        schedule.step(target=target, losses=losses)
    else:
        schedule.step()
```

(And we may want to deprecate the `_ScheduleForwardOnly`? open to discussion)

cc awgu wanchaol fegin fduwjj wz337 wconstab d4l3k 

[ghstack-poisoned]

wconstab commented:
question on the approach:

is your goal to make it so that schedule.step() will behave differently under `no_grad()` and not run backwards?

if so, i wonder if such polymorphic behavior of .step() is really needed.  Should we instead just add an api like `eval()` to the Schedule class, and have this API be equivalent to running ScheduleForwardOnly?
* also, i would be in favor of having this convenience method on all the scheudles and getting rid of ScheduleForwardOnly class

H-Huang commented:
@wconstab 

> is your goal to make it so that schedule.step() will behave differently under no_grad() and not run backwards?

Yep, that's correct

> Should we instead just add an api like eval() to the Schedule class, and have this API be equivalent to running ScheduleForwardOnly?

This would also work. We could do this instead but the minor API difference is you would need replace all `step()` calls in your code with `eval()`. I figured that we also would want to support `no_grad()` since currently it will raise an exception like "[X] for chunk X has gradients None and is expecting to send gradients to stage X" if run under `no_grad()` so this is fixing two issues.

So I guess we have 3 options:
1) Support step with `torch.no_grad()` and add `eval()` API
2) Add `eval()` API, add a more user friendly error message in `step()` when using `torch.no_grad()`
3) Only support step with `torch.no_grad()` (current implementation)

Any preferences for these options?

H-Huang committed:
Update on "[PP] Allow schedules to run under torch.no_grad()"


This change allows PP schedules to run under `with torch.no_grad():`

## Context

Currently, you can run "Forward only" for a schedule in two ways:
1. Use a custom schedule `_ScheduleForwardOnly`
2. Do not pass in `loss_fn` in schedule constructor, and no backward computations will be executed.

However, this is still limiting because we may want to run forward through the pipeline / calculate the loss, but without backward, e.g. during validation. These changes allow for this.

```python
with torch.no_grad():
    if self.rank == 0:
        schedule.step(x)
    elif self.rank == self.world_size - 1:
        losses = []
        schedule.step(target=target, losses=losses)
    else:
        schedule.step()
```

(And we may want to deprecate the `_ScheduleForwardOnly`? open to discussion)

cc awgu wanchaol fegin fduwjj wz337 wconstab d4l3k 

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] Add eval() API to schedule"


This change allows PP schedules to run under `with torch.no_grad():`

## Context

Currently, you can run "Forward only" for a schedule in two ways:
1. Use a custom schedule `_ScheduleForwardOnly`
2. Do not pass in `loss_fn` in schedule constructor, and no backward computations will be executed.

However, this is still limiting because we may want to run forward through the pipeline / calculate the loss, but without backward, e.g. during validation. These changes allow for this.

```python
with torch.no_grad():
    if self.rank == 0:
        schedule.step(x)
    elif self.rank == self.world_size - 1:
        losses = []
        schedule.step(target=target, losses=losses)
    else:
        schedule.step()
```

(And we may want to deprecate the `_ScheduleForwardOnly`? open to discussion)

cc awgu wanchaol fegin fduwjj wz337 wconstab d4l3k 

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] Add eval() API to schedule"


These change add an `eval()` API to PP schedules

## Context

Currently, you can run "Forward only" for a schedule in two ways:
1. Use a custom schedule `_ScheduleForwardOnly`
2. Do not pass in `loss_fn` in schedule constructor, and no backward computations will be executed.

However, this is still limiting because we may want to run forward through the pipeline / calculate the loss, but without backward, e.g. during validation. These changes allow for this.

```python
if self.rank == 0:
    schedule.eval(x)
elif self.rank == self.world_size - 1:
    losses = []
    schedule.eval(target=target, losses=losses)
else:
    schedule.eval()
```

TODO:
- in later PRs, we will deprecate the `_ScheduleForwardOnly`

cc awgu wanchaol fegin fduwjj wz337 wconstab d4l3k 

[ghstack-poisoned]

wconstab commented on torch/distributed/pipelining/stage.py:
```diff
@@ -761,6 +760,10 @@ def backward_one_chunk(
         after the last backward.
         """
+        # skip backward computation if backward is not enabled
+        if not self.has_backward:
```
ok. i guess i'd still _prefer_ if we could keep eval and step separate and just share helpers/utils for microbatching and output merging.  But i agree this is a good reason to use step for now and we could do that later as a cleanup if we want.
